Assignment5/Q1.py

```diff
@@ -0,0 +1,150 @@
+        index = get_index_of_first_differing_char(adjacent_word1, adjacent_word2)
+        if index != -1:
+            graph[adjacent_word1[index]].add(adjacent_word2[index])
+
+    return graph, state
+
+
+def find_total_order_from_char(partial_order, state, char, total_order):
+    """This method is the recursive part of topological sort (Similar to DFS from vertex).
+
+        Args:
+            partial_order: a dictionary of sets, graph data structure corresponding to characters partial order
+            state: a dictionary of strings, where strings indicate state of the character ("vertex")
+                   white - the character ("vertex") has not been visited by the search yet
+                   grey - the character ("vertex") is being visited, recursive call on it is not finished yet
+                   black - the character ("vertex") has already been visited
+            char: a character, from which the search is performed
+            total_order: a singly linked list of characters, the total order (lexicographical order) "accumulated
+                         so far" the new character is always added at the front of the list.
+    """
+    # The character is in search now.
+    state[char] = "grey"
+
+    # Check all characters that are "next" in the partial ordering.
+    for following_char in partial_order[char]:
+        # Perform a search from this character if it has not been visited yet.
+        if state[following_char] == "white":
+            find_total_order_from_char(partial_order, state, following_char, total_order)
+        # If any searched characters are found here, we found a cycle in the order, i.e. it is not a correct ordering.
+        elif state[following_char] == "grey":
+            raise ValueError("The given dictionary is not sorted in lexicographical order!")
+        break
+
+    # The search from this character is finished.
+    state[char] = "black"
+
+    # Update the total order (alphabet).
+    total_order.insert_first(char)
+
+
+def find_ordered_alphabet(dictionary):
+    """This method finds one of the possible lexicographic orders of dictionary characters given an ordered dictionary
+       words.
+
+        Args:
+            dictionary: an array of strings, list of language words in lexicographical order
+
+        Returns:
+            a list of characters, one the the possible lexicographical orderings of dictionary characters.
+
+        Throws:
+            ValueError: if the passed dictionary is empty or the word order in it is not valid.
+    """
+    # Empty dictionary.
+    if len(dictionary) == 0:
+        raise ValueError("The given dictionary is empty!")
+
+    # Create the partial ordering of dictionary characters.
+    (partial_order, state) = create_corresponding_partial_order(dictionary)
+
+    # Initialize the singly linked list. It will contain the total order of characters "accumulated so far".
+    total_order = SinglyLinkedList.SinglyLinkedList()
```

Comment:
I think you should be able to do this with a regular list.
At most, you might want to construct the list backward (since `append` is cheaper than `insert`).